File: Python/Tree/M-1372. Longest ZigZag Path in a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode], accu_zigzag_len: int, direction: str):
            if node is None:
                return accu_zigzag_len - 1

            left_accu_zigzag_len, right_accu_zigzag_len = 1, 1
            if direction == "left":
                right_accu_zigzag_len = accu_zigzag_len + 1
            elif direction == "right":
                left_accu_zigzag_len = accu_zigzag_len + 1

            left_zigzag_len = dfs(node.left, left_accu_zigzag_len, "left")
            right_zigzag_len = dfs(node.right, right_accu_zigzag_len, "right")
            zigzag_len = max(left_zigzag_len, right_zigzag_len)

            return zigzag_len

        return dfs(root, 0, "")

File: Tree/M-1372. Longest ZigZag Path in a Binary Tree/M-1372. Longest ZigZag Path in a Binary Tree.ts
/** Definition for a binary tree node. */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function longestZigZag(root: TreeNode | null): number {
  const dfs = (
    node: TreeNode | null,
    accuZigZagLength: number,
    direction: string
  ): number => {
    if (!node) return accuZigZagLength - 1;

    let [lAccuZigZagLength, rAccuZigZagLength] = [1, 1];
    if (direction === 'left') rAccuZigZagLength = accuZigZagLength + 1;
    else if (direction === 'right') lAccuZigZagLength = accuZigZagLength + 1;

    const lZigZagLength = dfs(node.left, lAccuZigZagLength, 'left');
    const rZigZagLength = dfs(node.right, rAccuZigZagLength, 'right');
    const zigZagLength = Math.max(lZigZagLength, rZigZagLength);

    return zigZagLength;
  };
  return dfs(root, 0, '');
}

/** case 1 */
const case1Lvl5Nodes = [new TreeNode(1)];
const case1Lvl4Nodes = [
  new TreeNode(1, null, case1Lvl5Nodes[0]),
  new TreeNode(1),
];
const case1Lvl3Nodes = [new TreeNode(1, case1Lvl4Nodes[0], case1Lvl4Nodes[1])];
const case1Lvl2Nodes = [
  new TreeNode(1, null, case1Lvl3Nodes[0]),
  new TreeNode(1),
];
const case1RootNode = new TreeNode(1, case1Lvl2Nodes[0], case1Lvl2Nodes[1]);

console.log(longestZigZag(case1RootNode)); // should be 4

/** case 2 */
const case2Lvl6Nodes = [new TreeNode(1)];
const case2Lvl5Nodes = [new TreeNode(1, null, case2Lvl6Nodes[0])];
const case2Lvl4Nodes = [
  new TreeNode(1, null, case2Lvl5Nodes[0]),
  new TreeNode(1),
];
const case2Lvl3Nodes = [
  new TreeNode(1),
  new TreeNode(1, case2Lvl4Nodes[0], case2Lvl4Nodes[1]),
];
const case2Lvl2Nodes = [new TreeNode(1, case2Lvl3Nodes[0], case2Lvl3Nodes[1])];
const case2RootNode = new TreeNode(1, null, case2Lvl2Nodes[0]);

console.log(longestZigZag(case2RootNode)); // should be 4
